pkg/util/watch.go

```diff
@@ -409,10 +359,72 @@ func Watch[C WatchClient[L], L metav1.ListMetaAccessor, T any, P WatchObject[T]]
 		}
 	}()
 
-	return &store, nil
+	return store, nil
+}
+
+type ProcessEventArgs[T any, P WatchObject[T]] struct {
+	Name   string
+	UID    types.UID
+	Config WatchConfig
+	Event  watch.Event
+	Object P
 }
 
-// WatchStore provides an interface for getting information about a list of Ts using the event
+func ProcessEvent[T any, P WatchObject[T]](
+	store *WatchStore[T],
+	handlers WatchHandlerFuncs[P],
+	args ProcessEventArgs[T, P],
+) error {
+	switch args.Event.Type {
+	case watch.Added:
+		store.mutex.Lock()
+		defer store.mutex.Unlock()
+
+		if _, ok := store.objects[args.UID]; ok {
+			return fmt.Errorf(
+				"watch %s: received add event for object %s that we already have",
+				args.Config.LogName, args.Name,
+			)
+		}
+		store.objects[args.UID] = (*T)(args.Object)
+		handlers.AddFunc((*T)(args.Object), false)
+	case watch.Bookmark:
+		// Nothing to do, just serves to give us a new ResourceVersion.
+	case watch.Deleted:
+		// We're given the state of the object immediately before deletion, which
+		// *may* be different to what we currently have stored.
+		store.mutex.Lock()
+		defer store.mutex.Unlock()
+
+		old, ok := store.objects[args.UID]
+		if !ok {
+			return fmt.Errorf(
+				"watch %s: received delete event for object %s that's not present",
+				args.Config.LogName, args.Name,
+			)
+		}
+		handlers.UpdateFunc(old, (*T)(args.Object))
+		delete(store.objects, args.UID)
+		handlers.DeleteFunc((*T)(args.Object), false)
+	case watch.Modified:
+		old, ok := store.objects[args.UID]
+		if !ok {
+			return fmt.Errorf(
+				"watch %s: received update event for object %s that's not present",
+				args.Config.LogName, args.Name,
+			)
+		}
+		store.objects[args.UID] = (*T)(args.Object)
+		handlers.UpdateFunc(old, (*T)(args.Object))
+	case watch.Error:
+		panic(errors.New("unreachable code reached")) // handled above
```

Comment:
note `// handled above` is still handled in the original function. It's worth adding a note that `ProcessEvent` must not be called with an event with type `watch.Error`. Should also change this panic message.